src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs

```diff
@@ -301,7 +314,32 @@ public IDisposable BeginScope<TState>(TState state)
                 throw new ArgumentNullException(nameof(state));
             }
 
-            return DictionaryLoggerScope.Push(state);
+            var stateValues = state as IDictionary<string, object>;
+            if (stateValues == null)
+            {
+                // There's nothing we can do without a dictionary.
+                return null;
+            }
+
+            string functionName = stateValues.GetValueOrDefault<string>(ScopeKeys.FunctionName);
+            string functionInvocationId = stateValues.GetValueOrDefault<string>(ScopeKeys.FunctionInvocationId);
+            string eventName = stateValues.GetValueOrDefault<string>(ScopeKeys.Event);
+
+            // If we have the invocation id, function name, and event, we know it's a new function. That means
+            // that we want to start a new operation and let App Insights track it for us.
+            if (!string.IsNullOrEmpty(functionName) &&
+                !string.IsNullOrEmpty(functionInvocationId) &&
+                eventName == LogConstants.FunctionStartEvent)
+            {
+                RequestTelemetry request = new RequestTelemetry();
```

Comment:
Actually, since we're dealing with the `ILogger` interface here, there's really no way to jump in and do something specific like this. However, in `BeginScope`, I did separate things out in the last commit so maybe that makes it a little easier to follow?

Note -- this complication is exactly the reason I didn't implement it this way in the first place. It's easier in the `ILogger` world to handle this as a 1-part operation and create the telemetry ourselves after the invocation is done. Breaking it into multiple parts complicates things, as you can see. But the benefits of getting 'free' Profiler support are worth it.